File: base/src/main/java/com/pichs/base/utils/MetaUtils.java
package com.pichs.base.utils;

import android.content.Context;
import android.content.pm.ApplicationInfo;
import android.content.pm.PackageManager;

/**
 * 清单文件中的 meta 数据获取
 */
public class MetaUtils {

    public static String getMetaDataString(Context context, String metaDataName, String... defaultValue) {
        String value = "";
        try {
            ApplicationInfo appInfo = context.getPackageManager().getApplicationInfo(
                    context.getPackageName(), PackageManager.GET_META_DATA);
            value = (defaultValue != null && defaultValue.length > 0) ? defaultValue[0] : value;
            value = appInfo.metaData.getString(metaDataName, value);
        } catch (PackageManager.NameNotFoundException e) {
            e.printStackTrace();
        }
        return value;
    }

    public static int getMetaDataInt(Context context, String metaDataName, int... defaultValue) {
        int value = 0;
        try {
            ApplicationInfo appInfo = context.getPackageManager().getApplicationInfo(
                    context.getPackageName(), PackageManager.GET_META_DATA);
            value = (defaultValue != null && defaultValue.length > 0) ? defaultValue[0] : value;
            value = appInfo.metaData.getInt(metaDataName, value);
        } catch (PackageManager.NameNotFoundException e) {
            e.printStackTrace();
        }
        return value;
    }

    public static boolean getMetaDataBoolean(Context context, String metaDataName, boolean... defaultValue) {
        boolean value = false;
        try {
            ApplicationInfo appInfo = context.getPackageManager().getApplicationInfo(
                    context.getPackageName(), PackageManager.GET_META_DATA);
            value = (defaultValue != null && defaultValue.length > 0) ? defaultValue[0] : value;
            value = appInfo.metaData.getBoolean(metaDataName, value);
        } catch (PackageManager.NameNotFoundException e) {
            e.printStackTrace();
        }
        return value;
    }

    public static Object getMetaData(Context context, String metaDataName) throws Exception {
        ApplicationInfo appInfo = context.getPackageManager().getApplicationInfo(
                context.getPackageName(), PackageManager.GET_META_DATA);
        return appInfo.metaData.get(metaDataName);
    }

    public static float getMetaDataFloat(Context context, String metaDataName, float... defaultValue) {
        float value = 0.0f;
        try {
            value = (float) getMetaData(context, metaDataName);
        } catch (Exception e) {
            value = (defaultValue != null && defaultValue.length > 0) ? defaultValue[0] : value;
        }
        return value;
    }

    @SafeVarargs
    public static <T> T getMetaData(Context context, String metaDataName, T... defaultValue) {
        T value = null;
        try {
            value = (T) getMetaData(context, metaDataName);
        } catch (Exception e) {
            value = (defaultValue != null && defaultValue.length > 0) ? defaultValue[0] : value;
        }
        return value;
    }

}

File: base/src/main/java/com/pichs/xbase/utils/MetaUtils.kt
package com.pichs.xbase.utils

import android.content.Context
import android.content.pm.PackageManager
import java.lang.Exception

/**
 * 清单文件中的 meta 数据获取
 */
object MetaUtils {
    fun getMetaDataString(context: Context, metaDataName: String, defaultValue: String? = null): String? {
        return try {
            val appInfo = context.packageManager.getApplicationInfo(
                context.packageName, PackageManager.GET_META_DATA
            )
            appInfo.metaData.getString(metaDataName, defaultValue) ?: defaultValue
        } catch (e: PackageManager.NameNotFoundException) {
            e.printStackTrace()
            defaultValue
        }
    }

    fun getMetaDataInt(context: Context, metaDataName: String, defaultValue: Int = 0): Int {
        return try {
            val appInfo = context.packageManager.getApplicationInfo(
                context.packageName, PackageManager.GET_META_DATA
            )
            val value = appInfo.metaData.getInt(metaDataName, defaultValue)
            value ?: defaultValue
        } catch (e: PackageManager.NameNotFoundException) {
            e.printStackTrace()
            defaultValue
        }
    }

    fun getMetaDataBoolean(context: Context, metaDataName: String, defaultValue: Boolean = false): Boolean {
        return try {
            val appInfo = context.packageManager.getApplicationInfo(
                context.packageName, PackageManager.GET_META_DATA
            )
            appInfo.metaData.getBoolean(metaDataName, defaultValue)
        } catch (e: PackageManager.NameNotFoundException) {
            e.printStackTrace()
            defaultValue
        }
    }

    fun getMetaDataFloat(context: Context, metaDataName: String, defaultValue: Float = 0.0f): Float {
        return try {
            getMetaData(context, metaDataName, defaultValue) ?: defaultValue
        } catch (e: Exception) {
            defaultValue
        }
    }

    fun <T> getMetaData(context: Context, metaDataName: String?, defaultValue: T? = null): T? {
        return try {
            getMetaData(context, metaDataName, defaultValue)
        } catch (e: Exception) {
            defaultValue
        }
    }
}